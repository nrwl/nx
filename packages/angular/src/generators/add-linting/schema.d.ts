export interface AddLintingGeneratorSchema {
  projectName: string;
  projectRoot: string;
  prefix: string;
  setParserOptionsProject?: boolean;
  skipFormat?: boolean;
  skipPackageJson?: boolean;
  unitTestRunner?: string;
  addPlugin?: boolean;
}
