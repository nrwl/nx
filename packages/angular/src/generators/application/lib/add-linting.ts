import type { Tree } from '@nx/devkit';
import type { NormalizedSchema } from './normalized-schema';

import { Linter } from '@nx/eslint';
import addLintingGenerator from '../../add-linting/add-linting';

export async function addLinting(host: Tree, options: NormalizedSchema) {
  if (options.linter === Linter.None) {
    return;
  }
  await addLintingGenerator(host, {
    projectName: options.name,
    projectRoot: options.appProjectRoot,
    prefix: options.prefix,
    setParserOptionsProject: options.setParserOptionsProject,
    skipPackageJson: options.skipPackageJson,
    unitTestRunner: options.unitTestRunner,
    skipFormat: true,
    addPlugin: options.addPlugin,
  });
}
