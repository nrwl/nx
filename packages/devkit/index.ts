/**
 * The Nx Devkit is the underlying technology used to customize Nx to support
 * different technologies and custom use-cases. It contains many utility
 * functions for reading and writing files, updating configuration,
 * working with Abstract Syntax Trees(ASTs), and more.
 *
 * As with most things in Nx, the core of Nx Devkit is very simple.
 * It only uses language primitives and immutable objects
 * (the tree being the only exception).
 *
 * @module @nx/devkit
 */

// eslint-disable-next-line @typescript-eslint/no-restricted-imports
export * from 'nx/src/devkit-exports';
export * from './public-api';
