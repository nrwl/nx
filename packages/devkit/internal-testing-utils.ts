/* eslint-disable @nx/enforce-module-boundaries */
export * from '../nx/src/internal-testing-utils/assert-valid-migrations';
export * from '../nx/src/internal-testing-utils/run-migration-against-this-workspace';
export * from '../nx/src/internal-testing-utils/with-environment';
export * from '../nx/src/internal-testing-utils/temp-fs';
/* eslint-enable @nx/enforce-module-boundaries */
export { setCwd } from './src/generators/artifact-name-and-directory-utils';
