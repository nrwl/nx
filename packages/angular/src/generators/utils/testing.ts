import type { Tree } from '@nx/devkit';
import { names, readProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Linter } from '@nx/eslint';
import { UnitTestRunner } from '../../utils/test-runners';
import { applicationGenerator } from '../application/application';
import type { Schema as ApplicationOptions } from '../application/schema';
import { componentGenerator } from '../component/component';
import { host } from '../host/host';
import type { Schema as HostOptions } from '../host/schema';
import { libraryGenerator } from '../library/library';
import type { Schema as LibraryOptions } from '../library/schema';
import { remote } from '../remote/remote';
import type { Schema as RemoteOptions } from '../remote/schema';

export async function generateTestApplication(
  tree: Tree,
  options: ApplicationOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await applicationGenerator(tree, {
    ...options,
  });
}

export async function generateTestHostApplication(
  tree: Tree,
  options: HostOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await host(tree, { ...options });
}

export async function generateTestRemoteApplication(
  tree: Tree,
  options: RemoteOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await remote(tree, { ...options });
}

export async function generateTestLibrary(
  tree: Tree,
  options: LibraryOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await libraryGenerator(tree, {
    ...options,
  });
}

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  tree.write('.gitignore', '');

  await libraryGenerator(tree, {
    directory: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleName: false,
    skipFormat: true,
    unitTestRunner: UnitTestRunner.Jest,
    standalone: false,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    path: `${libName}/src/lib/test-button/test-button`,
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  generateModule(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    path: `${libName}/src/lib/barrel/barrel-button/barrel-button`,
    module: 'barrel',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  generateModule(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    path: `${libName}/src/lib/variable-declare/variable-declare-button/variable-declare-button`,
    module: 'variable-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    path: `${libName}/src/lib/variable-declare/variable-declare-view/variable-declare-view`,
    module: 'variable-declare',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  generateModule(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-button/variable-spread-declare-button`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-view/variable-spread-declare-view`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-anotherview/variable-spread-declare-anotherview`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [
  VariableSpreadDeclareButtonComponent,
  VariableSpreadDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  generateModule(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    path: `${libName}/src/lib/static-member-declarations/cmp1/cmp1`,
    module: 'static-member-declarations',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    path: `${libName}/src/lib/static-member-declarations/cmp2/cmp2`,
    module: 'static-member-declarations',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  generateModule(tree, {
    name: 'nested',
    project: libName,
    path: `${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    module: 'nested',
    path: `${libName}/src/lib/nested/nested-button/nested-button`,
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    path: `${libName}/src/lib/test-other/test-other`,
    standalone: false,
    skipFormat: true,
  });

  return tree;
}

function generateModule(
  tree: Tree,
  options: { name: string; project: string; path?: string }
): void {
  const project = readProjectConfiguration(tree, options.project);

  if (options.path === undefined) {
    const sourceRoot = project.sourceRoot ?? `${project.root}/src`;
    const projectDirName =
      project.projectType === 'application' ? 'app' : 'lib';
    options.path = `${sourceRoot}/${projectDirName}`;
  }

  const moduleNames = names(options.name);
  const moduleFilePath = `${options.path}/${moduleNames.fileName}/${moduleNames.fileName}.module.ts`;

  tree.write(
    moduleFilePath,
    `import { NgModule } from '@angular/core';
  import { CommonModule } from '@angular/common';

  @NgModule({
    declarations: [],
    imports: [CommonModule],
  })
  export class ${moduleNames.className}Module {}`
  );
}
