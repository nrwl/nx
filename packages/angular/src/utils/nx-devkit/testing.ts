import type { Tree } from '@nx/devkit';
import { addProjectConfiguration, names, updateJson } from '@nx/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}

export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}

export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
${routing ? "import { RouterModule } from '@angular/router'" : ''};
import { App } from './app';
@NgModule({
  imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
  declarations: [App],
  bootstrap: [App]
})
export class AppModule {}
`
  );
  tree.write(
    `${appName}/src/main.ts`,
    `import { platformBrowser } from '@angular/platform-browser';

import { AppModule } from './app/app.module';

platformBrowser()
  .bootstrapModule(AppModule)
  .catch(err => console.log(err));
  `
  );
  tree.write(
    `${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  addProjectConfiguration(tree, appName, {
    root: `${appName}`,
    sourceRoot: `${appName}/src`,
    targets: {
      build: {
        options: {
          main: `${appName}/src/main.ts`,
        },
      },
      serve: {
        options: {},
      },
    },
    tags: [],
  });
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
@NgModule({
  imports: [
    CommonModule
  ],
  providers: []
})
export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `export * from './lib/${fileName}.module';
`
  );
}
