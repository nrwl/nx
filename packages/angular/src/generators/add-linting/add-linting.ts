import {
  formatFiles,
  joinPathFragments,
  runTasksInSerial,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { camelize, dasherize } from '@nx/devkit/src/utils/string-utils';
import { Linter, lintProjectGenerator } from '@nx/eslint';
import type * as eslint from 'eslint';
import {
  javaScriptOverride,
  typeScriptOverride,
} from '@nx/eslint/src/generators/init/global-eslint-config';
import {
  addOverrideToLintConfig,
  addPredefinedConfigToFlatLintConfig,
  findEslintFile,
  isEslintConfigSupported,
  replaceOverridesInLintConfig,
} from '@nx/eslint/src/generators/utils/eslint-file';
import { addAngularEsLintDependencies } from './lib/add-angular-eslint-dependencies';
import { isBuildableLibraryProject } from './lib/buildable-project';
import type { AddLintingGeneratorSchema } from './schema';
import { useFlatConfig } from '@nx/eslint/src/utils/flat-config';

export async function addLintingGenerator(
  tree: Tree,
  options: AddLintingGeneratorSchema
): Promise<GeneratorCallback> {
  const tasks: GeneratorCallback[] = [];
  const rootProject = options.projectRoot === '.' || options.projectRoot === '';
  const lintTask = await lintProjectGenerator(tree, {
    linter: Linter.EsLint,
    project: options.projectName,
    tsConfigPaths: [
      joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
    ],
    unitTestRunner: options.unitTestRunner,
    setParserOptionsProject: options.setParserOptionsProject,
    skipFormat: true,
    rootProject: rootProject,
    addPlugin: false,
    addExplicitTargets: true,
    skipPackageJson: options.skipPackageJson,
  });
  tasks.push(lintTask);

  if (isEslintConfigSupported(tree)) {
    const eslintFile = findEslintFile(tree, options.projectRoot);
    // keep parser options if they exist
    const hasParserOptions = tree
      .read(joinPathFragments(options.projectRoot, eslintFile), 'utf8')
      .includes(`${options.projectRoot}/tsconfig.*?.json`);

    if (useFlatConfig(tree)) {
      addPredefinedConfigToFlatLintConfig(
        tree,
        options.projectRoot,
        'flat/angular'
      );
      addPredefinedConfigToFlatLintConfig(
        tree,
        options.projectRoot,
        'flat/angular-template'
      );
      addOverrideToLintConfig(tree, options.projectRoot, {
        files: ['*.ts'],
        rules: {
          '@angular-eslint/directive-selector': [
            'error',
            {
              type: 'attribute',
              prefix: camelize(options.prefix),
              style: 'camelCase',
            },
          ],
          '@angular-eslint/component-selector': [
            'error',
            {
              type: 'element',
              prefix: dasherize(options.prefix),
              style: 'kebab-case',
            },
          ],
        },
      });
      addOverrideToLintConfig(tree, options.projectRoot, {
        files: ['*.html'],
        rules: {},
      });
    } else {
      replaceOverridesInLintConfig(tree, options.projectRoot, [
        ...(rootProject ? [typeScriptOverride, javaScriptOverride] : []),
        {
          files: ['*.ts'],
          ...(hasParserOptions
            ? {
                parserOptions: {
                  project: [`${options.projectRoot}/tsconfig.*?.json`],
                },
              }
            : {}),
          extends: [
            'plugin:@nx/angular',
            'plugin:@angular-eslint/template/process-inline-templates',
          ],
          rules: {
            '@angular-eslint/directive-selector': [
              'error',
              {
                type: 'attribute',
                prefix: camelize(options.prefix),
                style: 'camelCase',
              },
            ],
            '@angular-eslint/component-selector': [
              'error',
              {
                type: 'element',
                prefix: dasherize(options.prefix),
                style: 'kebab-case',
              },
            ],
          },
        },
        {
          files: ['*.html'],
          extends: ['plugin:@nx/angular-template'],
          /**
           * Having an empty rules object present makes it more obvious to the user where they would
           * extend things from if they needed to
           */
          rules: {},
        },
        ...(isBuildableLibraryProject(tree, options.projectName)
          ? [
              {
                files: ['*.json'],
                parser: 'jsonc-eslint-parser',
                rules: {
                  '@nx/dependency-checks': [
                    'error',
                    {
                      // With flat configs, we don't want to include imports in the eslint js/cjs/mjs files to be checked
                      ignoredFiles: [
                        '{projectRoot}/eslint.config.{js,cjs,mjs}',
                      ],
                    },
                  ],
                },
              } as any,
            ]
          : []),
      ]);
    }
  }

  if (!options.skipPackageJson) {
    const installTask = addAngularEsLintDependencies(tree, options.projectName);
    tasks.push(installTask);
  }

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return runTasksInSerial(...tasks);
}

export default addLintingGenerator;
