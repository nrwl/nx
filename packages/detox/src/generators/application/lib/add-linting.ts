import { Linter, lintProjectGenerator } from '@nx/eslint';
import {
  addDependenciesToPackageJson,
  GeneratorCallback,
  joinPathFragments,
  runTasksInSerial,
  Tree,
} from '@nx/devkit';
import { extraEslintDependencies } from '@nx/react';
import { NormalizedSchema } from './normalize-options';
import {
  addExtendsToLintConfig,
  addOverrideToLintConfig,
  addPredefinedConfigToFlatLintConfig,
  isEslintConfigSupported,
} from '@nx/eslint/src/generators/utils/eslint-file';
import { useFlatConfig } from '@nx/eslint/src/utils/flat-config';

export async function addLinting(host: Tree, options: NormalizedSchema) {
  if (options.linter === Linter.None) {
    return () => {};
  }

  const tasks: GeneratorCallback[] = [];
  const lintTask = await lintProjectGenerator(host, {
    linter: options.linter,
    project: options.e2eProjectName,
    tsConfigPaths: [
      joinPathFragments(options.e2eProjectRoot, 'tsconfig.app.json'),
    ],
    skipFormat: true,
    addPlugin: options.addPlugin,
  });
  tasks.push(lintTask);

  if (isEslintConfigSupported(host)) {
    if (useFlatConfig(host)) {
      addPredefinedConfigToFlatLintConfig(
        host,
        options.e2eProjectRoot,
        'flat/react'
      );
      // Add an empty rules object to users know how to add/override rules
      addOverrideToLintConfig(host, options.e2eProjectRoot, {
        files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
        rules: {},
      });
    } else {
      const addExtendsTask = addExtendsToLintConfig(
        host,
        options.e2eProjectRoot,
        { name: 'plugin:@nx/react', needCompatFixup: true }
      );
      tasks.push(addExtendsTask);
    }
  }

  const installTask = addDependenciesToPackageJson(
    host,
    extraEslintDependencies.dependencies,
    extraEslintDependencies.devDependencies
  );
  tasks.push(installTask);

  return runTasksInSerial(...tasks);
}
