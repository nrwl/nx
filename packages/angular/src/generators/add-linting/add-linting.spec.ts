import 'nx/src/internal-testing-utils/mock-project-graph';
import {
  ProjectConfiguration,
  Tree,
  addProjectConfiguration,
  readJson,
  updateJson,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import * as linter from '@nx/eslint';
import { addLintingGenerator } from './add-linting';

describe('addLinting generator', () => {
  let tree: Tree;
  const appProjectName = 'ng-app1';
  const appProjectRoot = `apps/${appProjectName}`;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    addProjectConfiguration(tree, appProjectName, {
      root: appProjectRoot,
      prefix: 'myOrg',
      projectType: 'application',
      targets: {},
    } as ProjectConfiguration);
  });

  it('should invoke the lintProjectGenerator', async () => {
    jest.spyOn(linter, 'lintProjectGenerator');

    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    expect(linter.lintProjectGenerator).toHaveBeenCalled();
  });

  it('should add the Angular specific EsLint devDependencies', async () => {
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const { devDependencies } = readJson(tree, 'package.json');
    expect(devDependencies['@angular-eslint/eslint-plugin']).toBeDefined();
    expect(
      devDependencies['@angular-eslint/eslint-plugin-template']
    ).toBeDefined();
    expect(devDependencies['@angular-eslint/template-parser']).toBeDefined();
  });

  it('should use flat config and install correct dependencies when using it', async () => {
    process.env.ESLINT_USE_FLAT_CONFIG = 'true';
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const { devDependencies } = readJson(tree, 'package.json');
    expect(devDependencies['@typescript-eslint/utils']).toMatchInlineSnapshot(
      `"^8.13.0"`
    );
    delete process.env.ESLINT_USE_FLAT_CONFIG;
  });

  it('should correctly generate the .eslintrc.json file', async () => {
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const eslintConfig = readJson(tree, `${appProjectRoot}/.eslintrc.json`);
    expect(eslintConfig).toMatchSnapshot();
  });

  it('should not touch the package.json when run with `--skipPackageJson`', async () => {
    let initialPackageJson;
    updateJson(tree, 'package.json', (json) => {
      json.dependencies = {};
      json.devDependencies = {};
      initialPackageJson = json;

      return json;
    });

    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
      skipPackageJson: true,
    });

    const packageJson = readJson(tree, 'package.json');
    expect(packageJson).toEqual(initialPackageJson);
  });
});
