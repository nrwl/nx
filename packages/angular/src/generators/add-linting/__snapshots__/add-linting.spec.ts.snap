// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addLinting generator should correctly generate the .eslintrc.json file 1`] = `
{
  "extends": [
    "../../.eslintrc.json",
  ],
  "ignorePatterns": [
    "!**/*",
  ],
  "overrides": [
    {
      "extends": [
        "plugin:@nx/angular",
        "plugin:@angular-eslint/template/process-inline-templates",
      ],
      "files": [
        "*.ts",
      ],
      "rules": {
        "@angular-eslint/component-selector": [
          "error",
          {
            "prefix": "my-org",
            "style": "kebab-case",
            "type": "element",
          },
        ],
        "@angular-eslint/directive-selector": [
          "error",
          {
            "prefix": "myOrg",
            "style": "camelCase",
            "type": "attribute",
          },
        ],
      },
    },
    {
      "extends": [
        "plugin:@nx/angular-template",
      ],
      "files": [
        "*.html",
      ],
      "rules": {},
    },
  ],
}
`;
